### PR TITLE
handle api errors in error boundary

### PR DESCRIPTION
**Pull request by selfcontained**
## Description

Adding a new `<QueryErrorBoundary>` to handle expected api errors that queries may throw. These will currently only come from the 2 queries we've enabled the `useErrorBoundary` option on.
* `getLoggedInUser()` - [code](https://github.com/gitpod-io/gitpod/blob/d7a546371fb2d1eb132ee1923270d8a18f7749a3/components/dashboard/src/hooks/use-user-loader.ts#L23)
* `listTeams()` - [code](https://github.com/gitpod-io/gitpod/blob/d7a546371fb2d1eb132ee1923270d8a18f7749a3/components/dashboard/src/data/organizations/orgs-query.ts#L31)

When these queries throw, they're caught first by the `<QueryErrorBoundary>`, which will handle any errors where `code` is one of `NOT_AUTHENTICATED` or `SETUP_REQUIRED` (we can add more in the future as we need). If it's not one of those errors, we re-throw it, and let it get caught by our default error boundary, now renamed to `<ReloadPageErrorBoundary>` for clarity. This one continues to report the error, and render a view to ask the user to reload the page. This should only be reached by unanticipated errors.

A nice side effect is it cleans up a bunch of the rendering in the main `<App/>` component.

## Related Issue(s)

Fixes WEB-70

## How to test

**Preview Env:** https://bmh-top-le869d9d8254.preview.gitpod-dev.com/workspaces
* On the preview environment, test logging in and out.

Other tests to exercise the error boundary are harder to test, as you have to simulate errors in different spots.
* Can throw different "errors" from the [user loader query](https://github.com/gitpod-io/gitpod/blob/d7a546371fb2d1eb132ee1923270d8a18f7749a3/components/dashboard/src/hooks/use-user-loader.ts#L28)
```ts
// simulate a not authenticated error
throw { code: 401, message: "simulated error" };
// simulate a needs setup error
throw { code: 410, message: "simulated error" };
// Simulate a server error
throw { code: 500, message: "simulated error" };
```

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+const Setup = lazy(() => import(/* webpackPrefetch: true */ "../../Setup"));
+
+// Error boundary intended to catch and handle expected errors from api calls
+export const QueryErrorBoundary: FC = ({ children }) => {
+    return (
+        <QueryErrorResetBoundary>
+            {({ reset }) => (
+                // Passing reset to onReset so any queries know to retry after boundary is reset
+                <ErrorBoundary FallbackComponent={ExpectedQueryErrorsFallback} onReset={reset}>
```
Comment (selfcontained) [inline components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx:25]: passing `reset` into here will tell react-query to reset any queries once the boundary has been reset.

File: components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        // Redirects if it's the root, no user, and no gp cookie present (has logged in recently)
+        if (isGitpodIo() && window.location.pathname === "/" && window.location.hash === "") {
+            // If there's no gp cookie, bounce to www site
+            if (!GitpodCookie.isPresent(document.cookie)) {
+                window.location.href = `https://www.gitpod.io`;
+                return <div></div>;
+            }
+        }
+
+        return <Login onLoggedIn={resetErrorBoundary} />;
```
Comment (selfcontained) [inline components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx:50]: It's important to reset the error boundary once we've gotten into a state that shouldn't produce it anymore. This tells react (and react query) to reset state, and try the render again.

File: components/dashboard/src/components/error-boundaries/ReloadPageErrorBoundary.tsx [hunk 3]
```diff
@@ -43,9 +48,9 @@ export const DefaultErrorFallback: FC<FallbackProps> = ({ error, resetErrorBound
                 .
             </Subheading>
             <div>
-                <button onClick={resetErrorBoundary}>Reload</button>
+                <button onClick={handleReset}>Reload</button>
```
Comment (selfcontained) [inline components/dashboard/src/components/error-boundaries/ReloadPageErrorBoundary.tsx:51]: I this case we don't actually want to ask React to try and re-render, we want to just have the user reload the page. We may want to have a route-level error boundary that does attempt a re-render on reset though.

File: components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx (hunk 1 above)
Comment (geropl) [reply components/dashboard/src/components/error-boundaries/QueryErrorBoundary.tsx:25]: Ah, nice! Wondered before how this happens/if it's the default/hard-wired behavior :bulb: :orange_heart:

File: components/dashboard/src/hooks/use-user-loader.ts [hunk 4]
```diff
@@ -4,55 +4,38 @@
...
-                    if (error.code >= 500) {
-                        throw error;
-                    }
-
-                    // Other errors will treat user as needing to log in
-                }
-            } finally {
-                trackLocation(!!user);
-            }
+            const user = await getGitpodService().server.getLoggedInUser();
```
Comment (geropl) [inline components/dashboard/src/hooks/use-user-loader.ts:23]: Wow. This looking really clean now! :tada:

Comment (geropl) [reply components/dashboard/src/hooks/use-user-loader.ts:23]: Actually, I think this in combination with the QueryErrorBoundary, could serve as one example on how to use react-query.
WDYT @selfcontained?

File: components/dashboard/src/App.tsx [hunk 5]
```diff
@@ -4,32 +4,23 @@
...
 import { useAnalyticsTracking } from "./hooks/use-analytics-tracking";
 import { useUserLoader } from "./hooks/use-user-loader";
 import { Login } from "./Login";
-import { isGitpodIo } from "./utils";
-
-const Setup = React.lazy(() => import(/* webpackPrefetch: true */ "./Setup"));
 
 // Wrap the App in an ErrorBoundary to catch User/Org loading errors
 // This will also catch any errors that happen to bubble all the way up to the top
 const AppWithErrorBoundary: FC = () => {
```
Comment (geropl) [inline components/dashboard/src/App.tsx:17]: This is superfluous now, right? Or am I missing sth? (asking to understand)

**Review by geropl — APPROVED** [review]
The code LGTM, tested and works. :heavy_check_mark: 

I have some minor clarification questions.

/hold But I think it's worth discussion if we want to keep main clean during the offsite, and have a "merge freeze" (see slack)

**Comment by geropl** [comment]
> /hold But I think it's worth discussion if we want to keep main clean during the offsite, and have a "merge freeze" (see slack)

*exhaling* roboquat... :disappointed:

**Pull request merged**